metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java

```diff
@@ -18,20 +18,39 @@
 package org.apache.kafka.metadata;
 
 import org.apache.kafka.common.Endpoint;
+import org.apache.kafka.common.Node;
 import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.metadata.RegisterBrokerRecord;
+import org.apache.kafka.common.metadata.RegisterBrokerRecord.BrokerEndpoint;
+import org.apache.kafka.common.metadata.RegisterBrokerRecord.BrokerFeature;
+import org.apache.kafka.common.security.auth.SecurityProtocol;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
 
+import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.stream.Collectors;
 
+import static org.apache.kafka.common.metadata.MetadataRecordType.REGISTER_BROKER_RECORD;
+
+
 /**
  * An immutable class which represents broker registrations.
  */
 public class BrokerRegistration {
+    private static Map<String, Endpoint> listenersToMap(Collection<Endpoint> listeners) {
+        Map<String, Endpoint> listenersMap = new HashMap<>();
+        for (Endpoint endpoint : listeners) {
+            listenersMap.put(endpoint.listenerName().get(), endpoint);
```

Comment:
I think we are guaranteed to have the listener present, but perhaps it's worth checking explicitly and throwing if it is not the case.

Reply:
Supposedly it's mandatory for brokers, but optional for clients (which really means they should have used a different type for the optional scenarios.)

I will add a check to the constructor that all listeners have names.